### PR TITLE
Aftershock: You can now pad clothes with neoprene to rainproof them

### PR DESCRIPTION
**Pull request by Ilysen**
#### Summary
```SUMMARY: Features "The tailor's kit can now pad clothes with neoprene to make them rainproof at a small encumbrance cost."```

#### Purpose of change
This was suggested on Discord and I thought it was an interesting change. The tailor's kit lacks interesting uses besides flat number increases and this helps to remedy that.

#### Describe the solution
You can use a tailor's kit and neoprene sheets to modify clothes to protect you from the rain, preventing you from getting wet.

Unlike clothing that's built as rainproofed, like the rain coat or survivor duster, this applies only to the covered body part; a rainproofed hat will only protect your head, and so on. You'll need to rainproof your head, torso, arms, and legs for full protection from the rain.

Because it doesn't provide a flat benefit like other mods, it has a lower encumbrance penalty, being the lowest of any of the mods. On average, it'll add around 2-3 encumbrance.

There's a new external option, `CLOTHING_WATERPROOFING`, that's disabled by default, to allow this feature. Aftershock has it enabled, but other mods, or json editing, can enable it freely.

#### Describe alternatives you've considered
Making more clothing rainproof. Things like Kevlar vests would presumably not soak up water. I decided to take this route because it allows more customization on the player's part.

#### Additional context
Tailor kit interface

### CODE REVIEW

File: src/iuse.cpp [hunk 1]
```diff
@@ -1549,14 +1549,15 @@ int iuse::sew_advanced( player *p, item *it, bool, const tripoint & )
...
+                          mod.item_tags.count( "neoprene_padded" );
 
     // We need extra thread to lose it on bad rolls
     const int thread_needed = mod.volume() / 125_ml + 10;
     // Returns true if the item already has the mod or if we have enough materials and thread to add it
     const auto can_add_mod = [&]( const std::string & new_mod, const itype_id & mat_item ) {
-        return mod.item_tags.count( new_mod ) > 0 ||
-               ( it->charges >= thread_needed && has_enough[mat_item] );
+        return ( mod.item_tags.count( new_mod ) > 0 ||
+                 ( it->charges >= thread_needed && has_enough[mat_item] ) && mod.get_coverage() >= 40 );
```
Comment (anothersimulacrum) [inline src/iuse.cpp]:
```c++
        return ( ( mod.item_tags.count( new_mod ) > 0 ||
                 ( it->charges >= thread_needed && has_enough[mat_item] ) ) && mod.get_coverage() >= 40 );
```
Think should fix your gorgon failure.

**Comment by Photoloss** [comment]
How many people would neoprene-pad glasses versus fur-lining their underwear? You are banning the latter even though it *actually makes sense* just to have a quick&dirty fix for potentially broken rainproofing mechanics.

**Comment by Ilysen** [comment]
I don't mind implementing some sort of fix, but the fact of the matter is that my knowledge of C++ is just too lacking and limited for me to do so effectively. I don't know how I'd do that, and believe me, I did try. When I get some time, I will either lower the threshold, or implement a json flag that prevents items from being modified.

**Comment by Ilysen** [comment]
@mlangsdorf gave me some help in Discord, and this now appears to work on a by-part, by-item basis. You can now keep padding your glasses with Kevlar and other silly things if you want.

**Comment by kevingranade** [comment]
My first impression is this needs to be a (very short) whitelist of clothes
rather than generally applicable.
My second is I can't think of ANY clothes where I would expect this to work.
Do you have any examples of anyone doing this kind of thing?

**Comment by Ilysen** [comment]
I haven't; my main motivation for making the change is a suggestion on Discord. Some research has revealed that some clothing waterproofing can be done with purchased spray, or, more likely to apply to the tailoring skill, vinyl used to laminate or create clothes. ([1](https://www.ehow.com/facts_4898704_what-vinyl-fabric.html)) ([2](https://www.sailrite.com/Waterproof-vs-Water-Resistant-Fabrics))

I'd be happy to implement this to replace the neoprene with vinyl lamination, though it would obviously be irreversible, and I'm not sure we *have* vinyl in the game. Rain coats typically have vinyl in them, and right now, those are just made of plastic; would it just utilize plastic chunks, or maybe a vinyl sheet craftable using plastic? Alternatively, would there be a new vinyl material that rain coats and rain hoods get, and that's where you'd find it?

**Comment by kevingranade** [comment]
The spray obviously needs to be found.
Harvesting from existing rainproof clothes seems like a reasonable place to get vinyl.

Either way, it needs to be a whitelist of clothes that can shed water properly if waterproofed, just spraying waterproofing on a shirt isn't going to accomplish much.

**Comment by ifreund** [comment]
Yeah I'm very skeptical of the effectiveness of aftermarket waterproofing for pretty much any clothing that wasn't originally intended to be waterproof.  You could definitely make large areas waterproof by adding waterproof material, but the seams/zippers/buttons will almost certainly leak without a high level of precision and some kind of seam sealing.  

In my experience vinyl is seldom, if ever used in rain clothing *because* it is completely waterproof.  If you made a fully enclosed jacket out of vinyl you would get soaked in your own sweat as it can't evaporate through the material.  Rain clothing is typically made of some kind of waterproof-breathable fabric.  REI has a pretty good overview of this technology [here](https://www.rei.com/learn/expert-advice/rainwear-how-it-works.html).  Basically, a survivor probably isn't going to get anywhere close to replicating high end rain gear.

The spray you mentioned applies what's known as a Durable Water Repellent (DWR) coating. As can be read on the REI page, it is often used in conjunction with a waterproof-breathable fabric.  It basically causes water to bead up and shed rather than soaking into the shell.  This keeps the breathability of the fabric high, as fabrics soaked with water are not breathable.  It can also be applied to non-waterproof fabrics with varying effectiveness.  It would cause a polyester windbreaker to be water resistant, but the coating would eventually "wet out."  I don't think this would have any effect if applied to a cotton t-shirt.

The other approach instead of using a waterproof-breathable fabric is to add lots of ventilation to the jacket (pit zips, back ventilation, etc). Still, Silnylon or Silpoly make much better choices than vinyl because of their reduced weight.  Here's an [example](https://antigravitygear.com/shop/product-category/ultralight-backpacking-rain-jacket-with-pit-zips/) of such a jacket.

Man I really wasn't planning on writing this much about rain jackets.

**Comment by Ilysen** [comment]
I was planning on giving certain clothes a `LAMINATES` flag that allows them to have full effectiveness in protecting against rain, and using vinyl on any other clothes would simply *partially* protect from the rain, instead of completely. For simplicity's sake, I planned on using vinyl as one type of fabric to represent such things, like how windbreakers, emergency jackets, and rain coats are currently made of plastic.

What do you think would be the best course of action to take? I'd really like to keep this idea in somehow while maintaining realism, but it's seeming more and more like it'll have to be modded in.

**Comment by ifreund** [comment]
Well, this discussion kinda highlights why we need [this](https://github.com/CleverRaven/Cataclysm-DDA/projects/25).  Implementing different fabric types is part of that and would solve this.

I'm not sure there is a good way to do this in the current system while holding to a high standard of realism.  If we just want to maintain the status quo when it comes to tailoring realism, a clothing whitelist and neoprene + duct tape as components would be fine imo.  Some things like survivor cargo pants are waterproofed with only 100 duct tape.

**Comment by Ilysen** [comment]
I decided that the best course of action for the feature without larger-scale overhauls was, for the time being, to make it a moddable function. I tied it to an external option called `CLOTHING_WATERPROOFING`, and set it to false by default.

Aftershock's `game_balance.json` has the value set to true, and other mods can set it to that if they'd like the feature as well.

**Comment by kevingranade** [comment]
So the status quo as I understand it is that adding neoprene patches as a waterproofing mechanism is not a thing that is done, ever. Why are you adding it, even as a mod?

Please look into restricting this further to your mod, for example by making the tailoring kit itself moddable, and by defining the new material (if needed) in your mod.

If neoprene patching is just a flag, what's stopping you from directly applying the waterproof flag instead?

**Comment by Ilysen** [comment]
To put it frankly, I'm adding it because I think it's an interesting mechanic. Aftershock already doesn't give a hoot about realism, and I don't care for maintaining realism that much in it either. If something makes sense in a narrow perspective without looking at it deeply (non-waterproof thing + waterproof sheet = waterproof thing!), then that's mostly all that matters to me when I add to Aftershock. I don't care about realism for Aftershock; I only care about balance and what's interesting.

Neoprene padding works in a certain way - that is, it protects that limb from rain only, instead of from all water. It's less useful than proper rainproof gear (since rainproof things like the rain coat protect the whole body) but still effective when applied to multiple pieces of gear.

But yeah, I could jsonize it, but I don't know how json is handled in the code. I'll need some time to work it out.

**Comment by Ilysen** [comment]
In the interest of time and getting rid of my bloated pull request list, I will work on jsonizing this at a later date. If it's acceptable to implement as-is, go ahead, but if not, feel free to close it.

**Comment by ZhilkinSerg** [comment]
You can always reopen PR later.

**Pull request closed without merging**